// infrastructurefactory/infrastructurefactory-aws-sharing.js
/*
	Factory for AWS Infrastructure SHARING

	The AWS sequence is:
		Source account: image-describe → get snapshotid
		- Source account: snapshot-copy-unencrypted → get unencrypted snapshot
		- OR
		- Source account: snapshot-copy-reencrypt → get unencrypted snapshot
		Source account: snapshot-describe → poll until state: 'completed'
		Source account: image-register → create AMI from unencrypted snapshot
		Source account: image-share → share AMI + snapshot with target account
		Target account: image-copy → copy with kmsKeyId to re-encrypt with your own key
		Target account: image-describe → poll until state: 'available'.

	---

	Then the updated workflow becomes:

		Source: Create custom KMS key
		Source: snapshot-copy-reencrypt → re-encrypt with custom key
		Source: snapshot-describe → poll until completed
		Source: image-register → create AMI from re-encrypted snapshot
		Source: Update custom KMS key policy to grant target account kms:Decrypt, kms:CreateGrant, kms:DescribeKey
		Source: image-share → share AMI + snapshot
		
		Target: image-copy → copy with target KMS key
		Target: image-describe → poll until available

	---

	Source Account:

	Get source account id:
	- lambda-local -l index.js -t 9000 -e events/lab/event-aws-sts-get-caller-identity-lab.json

	- lambda-local -l index.js -t 9000 -e events/lab/event-aws-ec2-image-describe-lab.json

	app-process-aws-ec2-snapshot-copy-reencrypt
		- lambda-local -l index.js -t 9000 -e events/lab/event-aws-ec2-snapshot-copy-reencrypt-lab.json
		- lambda-local -l index.js -t 9000 -e events/lab/event-aws-ec2-snapshot-copy-unencrypted-lab.json

	Check snapshot status:
	- lambda-local -l index.js -t 9000 -e events/lab/event-aws-ec2-snapshot-describe-lab.json

	Once snapshot available:
	- lambda-local -l index.js -t 9000 -e events/lab/event-aws-ec2-image-register-lab.json

	Update the encryption key used for the reencrypt to be used by the target account to decrypt.
	- lambda-local -l index.js -t 9000 -e events/lab/event-aws-kms-key-policy-update-lab.json

	Share the new image:
	- lambda-local -l index.js -t 9000 -e events/lab/event-aws-ec2-image-share-lab.json

	

*/

var entityos = require('entityos')
var _ = require('lodash')
var moment = require('moment');

module.exports = 
{
	VERSION: '0.0.1',

	init: function (param)
	{
		// Step 0:

		entityos.add(
		{
			name: 'app-process-aws-ec2-image-describe',
			code: function ()
			{
				const event = entityos.get({scope: '_event'});
				const imageIds = _.get(event, 'imageIds');

				if (imageIds == undefined)
				{
					entityos.invoke('util-end', 'No imageIds set', '400');
					return;
				}

				const commandParams = {
					ImageIds: _.isArray(imageIds) ? imageIds : [imageIds]
				};

				const infrastructureConfig = entityos.invoke('util-aws-get-config');
				const { EC2Client, DescribeImagesCommand } = require("@aws-sdk/client-ec2");
				const ec2 = new EC2Client(infrastructureConfig);

				const command = new DescribeImagesCommand(commandParams);

				ec2.send(command).then(function(response)
				{
					const responseData = _.map(response.Images, function(image)
					{
						return {
							imageId: image.ImageId,
							name: image.Name,
							state: image.State,
							stateReason: image.StateReason,
							blockDeviceMappings: image.BlockDeviceMappings
						};
					});

					entityos.invoke('util-end', responseData, '200');
				})
				.catch(function(error)
				{
					entityos.invoke('util-end', { error: error.message, code: error.Code }, '500');
				});
			}
		});

		// Step 1: Copy the encrypted snapshot as unencrypted (source account)
		entityos.add(
		{
			name: 'app-process-aws-ec2-snapshot-copy-unencrypted',
			code: function ()
			{
				const event = entityos.get({scope: '_event'});
				const snapshotId = _.get(event, 'snapshotId');
				const region = _.get(event, 'region', 'ap-southeast-2');
				const description = _.get(event, 'description', 'Unencrypted copy');

				if (snapshotId == undefined)
				{
					entityos.invoke('util-end', 'No snapshotId set', '400');
					return;
				}

				const infrastructureConfig = entityos.invoke('util-aws-get-config');
				const { EC2Client, CopySnapshotCommand } = require("@aws-sdk/client-ec2");
				const ec2 = new EC2Client(infrastructureConfig);

				const command = new CopySnapshotCommand({
					SourceSnapshotId: snapshotId,
					SourceRegion: region,
					Description: description,
					Encrypted: false
				});

				ec2.send(command).then(function(response)
				{
					entityos.invoke('util-end', {
						snapshotId: response.SnapshotId,
						status: 'copying'
					}, '200');
				})
				.catch(function(error)
				{
					entityos.invoke('util-end', { error: error.message, code: error.Code }, '500');
				});
			}
		});

		// Step 2: Check snapshot copy status
		entityos.add(
		{
			name: 'app-process-aws-ec2-snapshot-describe',
			code: function ()
			{
				const event = entityos.get({scope: '_event'});
				const snapshotIds = _.get(event, 'snapshotIds');

				if (snapshotIds == undefined)
				{
					entityos.invoke('util-end', 'No snapshotIds set', '400');
					return;
				}

				const infrastructureConfig = entityos.invoke('util-aws-get-config');
				const { EC2Client, DescribeSnapshotsCommand } = require("@aws-sdk/client-ec2");
				const ec2 = new EC2Client(infrastructureConfig);

				const command = new DescribeSnapshotsCommand({
					SnapshotIds: _.isArray(snapshotIds) ? snapshotIds : [snapshotIds]
				});

				ec2.send(command).then(function(response)
				{
					const responseData = _.map(response.Snapshots, function(snapshot)
					{
						return {
							snapshotId: snapshot.SnapshotId,
							state: snapshot.State,
							progress: snapshot.Progress,
							encrypted: snapshot.Encrypted,
							volumeSize: snapshot.VolumeSize,
							description: snapshot.Description
						};
					});

					entityos.invoke('util-end', responseData, '200');
				})
				.catch(function(error)
				{
					entityos.invoke('util-end', { error: error.message, code: error.Code }, '500');
				});
			}
		});

		// Step 3: Register a new AMI from the unencrypted snapshot
		entityos.add(
		{
			name: 'app-process-aws-ec2-image-register',
			code: function ()
			{
				const event = entityos.get({scope: '_event'});
				const register = _.get(event, 'register');

				if (register == undefined)
				{
					entityos.invoke('util-end', 'No register params set', '400');
					return;
				}

				const snapshotId = _.get(register, 'snapshotId');
				const name = _.get(register, 'name');
				const volumeSize = _.get(register, 'volumeSize', 100);
				const architecture = _.get(register, 'architecture', 'x86_64');
				const rootDeviceName = _.get(register, 'rootDeviceName', '/dev/sda1');

				if (snapshotId == undefined || name == undefined)
				{
					entityos.invoke('util-end', 'snapshotId and name are required', '400');
					return;
				}

				const infrastructureConfig = entityos.invoke('util-aws-get-config');
				const { EC2Client, RegisterImageCommand } = require("@aws-sdk/client-ec2");
				const ec2 = new EC2Client(infrastructureConfig);

				const command = new RegisterImageCommand({
					Name: name,
					Architecture: architecture,
					RootDeviceName: rootDeviceName,
					VirtualizationType: 'hvm',
					EnaSupport: true,
					BlockDeviceMappings: [
					{
						DeviceName: rootDeviceName,
						Ebs: {
							SnapshotId: snapshotId,
							VolumeSize: volumeSize,
							VolumeType: 'gp2',
							DeleteOnTermination: true
						}
					}]
				});

				ec2.send(command).then(function(response)
				{
					entityos.invoke('util-end', {
						imageId: response.ImageId
					}, '200');
				})
				.catch(function(error)
				{
					entityos.invoke('util-end', { error: error.message, code: error.Code }, '500');
				});
			}
		});

		// Step 4: Share the AMI and snapshot with target account
		entityos.add(
		{
			name: 'app-process-aws-ec2-image-share',
			code: function ()
			{
				const event = entityos.get({scope: '_event'});
				const share = _.get(event, 'share');

				if (share == undefined)
				{
					entityos.invoke('util-end', 'No share params set', '400');
					return;
				}

				const imageId = _.get(share, 'imageId');
				const snapshotId = _.get(share, 'snapshotId');
				const targetAccountId = _.get(share, 'targetAccountId');

				if (imageId == undefined || snapshotId == undefined || targetAccountId == undefined)
				{
					entityos.invoke('util-end', 'imageId, snapshotId and targetAccountId are required', '400');
					return;
				}

				const infrastructureConfig = entityos.invoke('util-aws-get-config');
				const { EC2Client, ModifyImageAttributeCommand, ModifySnapshotAttributeCommand } = require("@aws-sdk/client-ec2");
				const ec2 = new EC2Client(infrastructureConfig);

				const shareSnapshot = new ModifySnapshotAttributeCommand({
					SnapshotId: snapshotId,
					Attribute: 'createVolumePermission',
					OperationType: 'add',
					UserIds: [targetAccountId]
				});

				const shareImage = new ModifyImageAttributeCommand({
					ImageId: imageId,
					LaunchPermission: {
						Add: [{ UserId: targetAccountId }]
					}
				});

				ec2.send(shareSnapshot).then(function()
				{
					return ec2.send(shareImage);
				})
				.then(function()
				{
					entityos.invoke('util-end', {
						status: 'shared',
						imageId: imageId,
						snapshotId: snapshotId,
						targetAccountId: targetAccountId
					}, '200');
				})
				.catch(function(error)
				{
					entityos.invoke('util-end', { error: error.message, code: error.Code }, '500');
				});
			}
		});

		// Step 5: Copy the shared AMI into target account with encryption (target account)
		entityos.add(
		{
			name: 'app-process-aws-ec2-image-copy',
			code: function ()
			{
				const event = entityos.get({scope: '_event'});
				const copy = _.get(event, 'copy');

				if (copy == undefined)
				{
					entityos.invoke('util-end', 'No copy params set', '400');
					return;
				}

				const sourceImageId = _.get(copy, 'sourceImageId');
				const sourceRegion = _.get(copy, 'sourceRegion', 'ap-southeast-2');
				const name = _.get(copy, 'name');
				const kmsKeyId = _.get(copy, 'kmsKeyId');

				if (sourceImageId == undefined || name == undefined)
				{
					entityos.invoke('util-end', 'sourceImageId and name are required', '400');
					return;
				}

				const commandParams = {
					SourceImageId: sourceImageId,
					SourceRegion: sourceRegion,
					Name: name
				};

				if (kmsKeyId)
				{
					commandParams.Encrypted = true;
					commandParams.KmsKeyId = kmsKeyId;
				}

				const infrastructureConfig = entityos.invoke('util-aws-get-config');
				const { EC2Client, CopyImageCommand } = require("@aws-sdk/client-ec2");
				const ec2 = new EC2Client(infrastructureConfig);

				const command = new CopyImageCommand(commandParams);

				ec2.send(command).then(function(response)
				{
					entityos.invoke('util-end', {
						imageId: response.ImageId,
						status: 'copying'
					}, '200');
				})
				.catch(function(error)
				{
					entityos.invoke('util-end', { error: error.message, code: error.Code }, '500');
				});
			}
		});

		// OTHER STEPS

		entityos.add(
		{
			name: 'app-process-aws-ec2-snapshot-copy-reencrypt',
			code: function ()
			{
				const event = entityos.get({scope: '_event'});
				const snapshotId = _.get(event, 'snapshotId');
				const region = _.get(event, 'region', 'ap-southeast-2');
				const kmsKeyId = _.get(event, 'kmsKeyId');
				const description = _.get(event, 'description', 'Re-encrypted copy');

				if (snapshotId == undefined || kmsKeyId == undefined)
				{
					entityos.invoke('util-end', 'snapshotId and kmsKeyId are required', '400');
					return;
				}

				const infrastructureConfig = entityos.invoke('util-aws-get-config');
				const { EC2Client, CopySnapshotCommand } = require("@aws-sdk/client-ec2");
				const ec2 = new EC2Client(infrastructureConfig);

				const command = new CopySnapshotCommand({
					SourceSnapshotId: snapshotId,
					SourceRegion: region,
					Description: description,
					Encrypted: true,
					KmsKeyId: kmsKeyId
				});

				ec2.send(command).then(function(response)
				{
					entityos.invoke('util-end', {
						snapshotId: response.SnapshotId,
						status: 'copying'
					}, '200');
				})
				.catch(function(error)
				{
					entityos.invoke('util-end', { error: error.message, code: error.Code }, '500');
				});
			}
		});

		entityos.add(
		{
			name: 'app-process-aws-kms-key-policy-update',
			code: function ()
			{
				const event = entityos.get({scope: '_event'});
				const keyId = _.get(event, 'keyId');
				const targetAccountId = _.get(event, 'targetAccountId');
				const sourceAccountId = _.get(event, 'sourceAccountId');

				if (keyId == undefined || targetAccountId == undefined || sourceAccountId == undefined)
				{
					entityos.invoke('util-end', 'keyId, sourceAccountId and targetAccountId are required', '400');
					return;
				}

				const infrastructureConfig = entityos.invoke('util-aws-get-config');
				const { KMSClient, PutKeyPolicyCommand } = require("@aws-sdk/client-kms");
				const kms = new KMSClient(infrastructureConfig);

				const policy = JSON.stringify({
					Version: "2012-10-17",
					Statement: [
						{
							Sid: "EnableRootAccountAccess",
							Effect: "Allow",
							Principal: {
								AWS: "arn:aws:iam::" + sourceAccountId + ":root"
							},
							Action: "kms:*",
							Resource: "*"
						},
						{
							Sid: "AllowTargetAccountAccess",
							Effect: "Allow",
							Principal: {
								AWS: "arn:aws:iam::" + targetAccountId + ":root"
							},
							Action: [
								"kms:Decrypt",
								"kms:CreateGrant",
								"kms:DescribeKey"
							],
							Resource: "*"
						}
					]
				});

				const command = new PutKeyPolicyCommand({
					KeyId: keyId,
					PolicyName: "default",
					Policy: policy
				});

				kms.send(command).then(function()
				{
					entityos.invoke('util-end', {
						status: 'updated',
						keyId: keyId,
						targetAccountId: targetAccountId
					}, '200');
				})
				.catch(function(error)
				{
					entityos.invoke('util-end', { error: error.message, code: error.Code }, '500');
				});
			}
		});

		// UTIL

		entityos.add(
		{
			name: 'app-process-aws-sts-get-caller-identity',
			code: function ()
			{
				const infrastructureConfig = entityos.invoke('util-aws-get-config');
				const { STSClient, GetCallerIdentityCommand } = require("@aws-sdk/client-sts");
				const sts = new STSClient(infrastructureConfig);

				const command = new GetCallerIdentityCommand({});

				sts.send(command).then(function(response)
				{
					entityos.invoke('util-end', {
						accountId: response.Account,
						arn: response.Arn,
						userId: response.UserId
					}, '200');
				})
				.catch(function(error)
				{
					entityos.invoke('util-end', { error: error.message, code: error.Code }, '500');
				});
			}
});
	}
}
